man/jpegpixi.1.in
.\" This file is part of jpegpixi, a program to interpolate pixels in
.\" JFIF image files.
.\" Copyright (C) 2002, 2003, 2004, 2005 Martin Dickopp
.
.\" Jpegpixi is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published
.\" by the Free Software Foundation; either version 2 of the License,
.\" or (at your option) any later version.
.
.\" Jpegpixi is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.
.\" You should have received a copy of the GNU General Public License
.\" along with jpegpixi; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301,
.\" USA.
.
.TH JPEGPIXI 1 20-Nov-2005 "@PACKAGE_STRING@" "User Commands"
.SH NAME
jpegpixi \- interpolate pixels in JFIF (JPEG) image files
.SH SYNOPSIS
.B jpegpixi
.RI [ OPTION ]...\&
.I SOURCE DEST
[[\fID\fR:]\fIX\fR,\fIY\fR[,\fIS\fR]|[,\fISX\fR,\fISY\fR]]...
.SH DESCRIPTION
Jpegpixi interpolates pixels or pixel blocks in JFIF images files
(commonly referred to as "JPEG images"). This is useful to correct images
from a digital camera with CCD defects. Only DCT blocks which contain
pixels to be interpolated are decoded and re-encoded, and the re-encoding
is performed with exactly the same parameters with which the image has
originally been encoded. Therefore, the image quality is preserved as much
as possible.
.PP
.I SOURCE
and
.I DEST
are the filenames of the source file and the destination file,
respectively. In both cases,
.B \-
can be used, which refers to standard input or standard output.
.PP
The destination filename is followed by any number of pixel or pixel block
specifications. It starts with an optional direction specifier
.RI ( D ),
which can be
.B 2
for 2-dimensional interpolation (the default),
.B V
or
.B v
for 1-dimensional vertical interpolation (e.g.\ to remove horizontal
stripes), or
.B H
or
.B h
for 1-dimensional horizontal interpolation.
.IR X , Y
are the coordinates of the pixel or the upper-left corner of the pixel
block.
.I S
is the size of the pixel block (1 by default). Alternatively,
separate sizes can be specified for the width
.RI ( SX )
and the height
.RI ( SY ).
.PP
All numbers
.RI ( X ,
.IR Y ,
.IR S ,
.IR SX ,
.IR SY )
can be expressed as absolute coordinates/sizes or percentages of the image
size. If a number is followed by a percent character
.RB ( % ),
it is interpreted as a percentage, otherwise as an absolute number.
.SH OPTIONS
.IP "\fB\-f\fR \fIFILE\fR, \fB\-\-blocks\-file\fR=\fIFILE\fR"
Read pixel block specifications from file
.I FILE
in addition to the command line. The file should contain one specification
([\fID\fR:]\fIX\fR,\fIY\fR[,\fIS\fR]|[,\fISX\fR,\fISY\fR]) per line. Empty
lines are ignored. The file may also contain comments, which start with an
octothorpe
.RB ( # )
and extend to the end of the line.
.IP "\fB\-m\fR \fIMETHOD\fR, \fB\-\-method\fR=\fIMETHOD\fR"
Use interpolation method
.I METHOD
(default:
.BR linear ).
See section INTERPOLATION METHODS below.
.IP "\fB\-v\fR, \fB\-\-verbose\fR"
Display the coordinates and size of each pixel block that is
interpolated.
.IP "\fB\-i\fR, \fB\-\-info\fR"
Display information about the image, such as the size of the image,
the colorspace in which it is encoded, and sampling rates for the
different components.
.IP "\fB\-s\fR, \fB\-\-strip\fR"
Do not copy comment and extra markers from the source to the
destination file. This strips preview images, EXIF data, and similar
information.
.IP "\fB\-\-help\fR"
Display a short help text and exit immediately.
.IP "\fB\-\-version\fR"
Display version information and exit immediately.
.SH "INTERPOLATION METHODS"
.IP "\fB0\fR, \fBav\fR, \fBaverage\fR"
The pixels adjacent to the pixel block are averaged. The resulting color is
assigned to all pixels in the block. For 1-dimensional interpolation, this
is done separately for one pixel wide, horizontal or vertical stripes.
.IP "\fB1\fR, \fBli\fR, \fBlinear\fR"
The pixels which have a distance of 1 from the pixel block are used to
calculate a bilinear surface (2-dim), or a group of linear curves (1-dim),
which is then used to assign colors to the pixels in the block.
.IP "\fB2\fR, \fBqu\fR, \fBquadratic\fR"
The pixels which have a distance of 2 or less from the pixel block are used
to calculate a biquadratic surface (2-dim), or a group of quadratic curves
(1-dim), which is then used to assign colors to the pixels in the block.
.IP "\fB3\fR, \fBcu\fR, \fBcubic\fR"
The pixels which have a distance of 3 or less from the pixel block are used
to calculate a bicubic surface (2-dim), or a group of cubic curves (1-dim),
which is then used to assign colors to the pixels in the block.
.SH BUGS
Please report bugs to <martin\-jpegpixi@zero\-based.org>.
.SH COPYRIGHT
Copyright \(co 2002, 2003, 2004, 2005 Martin Dickopp
.PP
Jpegpixi is free software; it may be copied and/or modified under the
terms of the GNU General Public License version 2 or (at your option)
any later version. There is NO warranty; not even for MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR jpeghotp (1)
.PP
The jpegpixi homepage:
http://www.zero\-based.org/software/jpegpixi/
.
.\" Local Variables:
.\" mode: nroff
.\" coding: iso-8859-1
.\" End:


NEWS
Version 1.1.1:
* Fixed typo in English jpegpixi manual page.
* Updated postal address of the Free Software Foundation.

Version 1.1.0:
* In addition to hot pixels in an otherwise black image, the `jpeghotp'
  program can now also find dead pixels in an otherwise white image.
* Fail gracefully if the locale cannot be set.

Version 1.0.3:
* Stephan Peccini <stephan@tesenca.info> has contributed French
  translations of the manual pages and the message catalog. Thank
  you, Stephan!

Version 1.0.2:
* Spelling mistakes in the German message catalog have been corrected.
* Minor cleanup of the code and build infrastructure. Split some
  messages to make the job of translators easier.

Version 1.0.1:
* Jarno Elonen <elonen@iki.fi> has contributed Finnish translations of
  the manual pages and the message catalog. Thank you, Jarno!

Version 1.0.0:
* Native Language Support is provided. The package contains a message
  catalog with German translations.
* With Native Language Support, all intended features are implemented.
  Therefore, the major version number is changed to 1.

Version 0.16.0:
* German translations of the manual pages are provided. This is the
  first step towards Native Language Support in the programs themselves.

Version 0.15.1:
* Following a proposal on the debian-legal mailinglist (see
  <http://lists.debian.org/debian-legal/2004/05/msg00150.html> and
  followups), license information (i.e. the output shown if the
  `--version' command line option is specified) is put into a separate
  section `.license' in the executable, if possible.  The contents of
  this section can be dumped with the GNU binutils, using the command:
  objcopy -O binary -j .license /path/to/jpegpixi `tty`
* Bug fix: Link with math library (-lm) if necessary.

Version 0.15.0:
* The `jpegpixi' program has a new command line option `--verbose' to
display coordinates and sizes of interpolated pixel blocks.
* Bug fix: Warn if overlapping pixel blocks are specified.
* Bug fix: Consolidate adjacent pixel blocks. For example, if `42,17'
and `43,17' are specified, the behavior is the same as if `42,17,2,1'
had been specified instead.
* Bug fix: Work correctly in extreme cases where almost the whole image
is interpolated.

Version 0.14.2:
* The `jpeghotp' program can output comments with the luminosities of
the hot pixel blocks. This can be helpful in "hand tuning" the pixel
blocks file.

Version 0.14.1:
* Jarno Elonen <elonen@iki.fi> will maintain the jpegpixi Debian package;
therefore the `debian' directory has been removed.

Version 0.14:
* A new program `jpeghotp' is provided. It finds hot pixels in an
otherwise black JPEG image.
* Spelling mistakes in the manual page corrected.

Version 0.13:
* Coordinates and sizes can not only be specified as absolute values, but
also as percentages of the image size.

Version 0.12:
* Jpegpixi now parses command line options with code generated with `opag',
an option parser generator by the same author. This has no user-visible effect.

Version 0.11:
* No code changes at all. The only purpose of this release is to incorporate
a change of the home page and the author's e-mail address.

Version 0.10:
* Pixel block specifications can be read from a file (in addition to
the command line).

Version 0.9:
* In addition to 2-dimensional interpolation, 1-dimensional interpolation is
now also supported. The latter can be used to remove horizontal or vertical
stripes which span the whole image or a large part of it. Thanks to Evgeny
Stambulchik for suggesting this.
* Bug fixes.

Version 0.8:
* The size of the block to be interpolated is now only limited by available
memory.
* Jpegpixi does no longer use precalculated tables for the interpolation;
this reduces the size of the executable.
* Especially for low-quality images, the (inevitable) quality loss in the
vicinity of interpolated pixels is now even smaller. To achieve this, the
optimal threshold for the rounding of DCT coefficients has been empirically
determined.
* Huffman tables are re-calculated for the output image, resulting in a
smaller file size in some cases.
* It is no longer possible to print debugging output. Maintaining this
functionality is too much of a hassle for too little benefit.

Version 0.7:
* Bug fix: Pixels near the image edges are treated correctly.
* More verbose debugging output.
* Uses automake 1.7.5.

Version 0.6:
* The package uses autoconf 2.57 and automake 1.7.2.
* A `debian' subdirectory to build a Debian package is provided.

Version 0.5:
* There is now a manual page.
* The maximum size of the pixel block to be interpolated has been
increased to 5x5 pixels.
* The package uses automake 1.6.1.

Version 0.4:
* Biquadratic and bicubic interpolation are supported in addition to
averaging and bilinear interpolation.

Version 0.3:
* The size of the pixel block to be interpolated can be specified
(up to 3x3 pixels).
* Interpolation engine rewritten from scratch. Averaging and bilinear
interpolation are available as interpolation methods. Quadratic
interpolation has been taken out and will be added again to a later
version. CAUTION: The names of the interpolation methods (`-m' option)
have changed.

Version 0.2:
* Bug fixes. Now jpegpixi works reasonably well with low quality JFIF
images.
* New command line option `-d' to display debugging information (e.g.
the DCT coefficients before and after the interpolation).
* Interpolation method can be specified on the command line. The
following methods are supported:
  - average of 4 adjacent pixels
  - average of 8 surrounding pixels
  - quadratic surface defined by 12 nearest pixels

Version 0.1:
* First release.


jpegpixi.c
/* This file is part of jpegpixi, a program to interpolate pixels in
   JFIF image files.
   Copyright (C) 2002, 2003, 2004, 2005 Martin Dickopp

   Jpegpixi is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published
   by the Free Software Foundation; either version 2 of the License,
   or (at your option) any later version.

   Jpegpixi is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with jpegpixi; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301,
   USA.  */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif
#include "util.h"

#include <stdio.h>
#if HAVE_ERRNO_H
# include <errno.h>
#endif
#ifndef errno
extern int errno;
#endif

#if STDC_HEADERS
# include <ctype.h>
#else
# define isspace(c) ((c) == ' ' || (c) == '\t')
# define isdigit(c) ((c) >= '0' && (c) <= '9')
#endif

#if HAVE_NLS
# include <locale.h>
#endif

#include "jpegpixi.h"
#include "rbtree.h"
#include "optpixi.h"


/* Name of this program.  */
#define PROGRAM_NAME PACKAGE_NAME



/* Display text in response to the --help command line option.  */
extern void display_help_text (void);

/* Display text in response to the --version command line option.  */
extern void display_version_text (void);



static char *read_line (FILE *f, const char *filename, char **buffer_ptr, size_t *len_buffer_ptr,
                        unsigned int *lineno_ptr) gcc_attr_nonnull (());
static struct point_dimdir_s *parse_struct_point_dimdir_s (struct point_dimdir_s *point,
                                                           const char *str, unsigned int image_width,
                                                           unsigned int image_height) gcc_attr_nonnull (());
static int point_dimdir_s_validate_and_consolidate (void *a, const void *b) gcc_attr_nonnull (());



int
main (int argc, char *argv [])
{
    int i;
    enum method_t method = linear;
    const char *infilename, *outfilename;
    unsigned int image_width, image_height;
    struct rbtree points;



    /* Store program invocation name.  */
    invocation_name = argv [0] != 0 && *(argv [0]) != '\0' ? argv [0] : PROGRAM_NAME;


    /* Set up Native Language Support.  */
#if HAVE_NLS
    if (setlocale (LC_ALL, "") != 0)
    {
        if (bindtextdomain (PACKAGE_TARNAME, LOCALEDIR) == 0 || textdomain (PACKAGE_TARNAME) == 0)
            mem_alloc_failed ();
    }
    else
    {
        fprintf (stderr, "%s: cannot set locale\n"
                 "Make sure the `LC_*' and `LANG' environment variables are set to valid values.\n", invocation_name);
        exit (1);
    }
#endif


    /* Parse command line options.  */
    if ((i = parse_options (invocation_name, argc, argv)) == -1)
        goto error_try_help;

    if (opt_help)
    {
        display_help_text ();
        exit (0);
    }

    if (opt_version)
    {
        display_version_text ();
        exit (0);
    }

    if (opt_method)
    {
        if (strcmp ("0", arg_method) == 0 || strcmp ("av", arg_method) == 0
            || strcmp ("average", arg_method) == 0)
            method = average;
        else if (strcmp ("1", arg_method) == 0 || strcmp ("li", arg_method) == 0
                 || strcmp ("linear", arg_method) == 0)
            method = linear;
        else if (strcmp ("2", arg_method) == 0 || strcmp ("qu", arg_method) == 0
                 || strcmp ("quadratic", arg_method) == 0)
            method = quadratic;
        else if (strcmp ("3", arg_method) == 0 || strcmp ("cu", arg_method) == 0
                 || strcmp ("cubic", arg_method) == 0)
            method = cubic;
        else
        {
            fprintf (stderr, _("%s: invalid interpolation method `%s'\n"), invocation_name, arg_method);
            goto error_try_help;
        }
    }


    /* Read filenames from command line.  */
    if (i < argc)
        infilename = argv [i++];
    else
    {
        fprintf (stderr, _("%s: source filename missing\n"), invocation_name);
      error_try_help:
        fprintf (stderr, _("Try `%s --help' for more information.\n"), invocation_name);
        exit (1);
    }

    if (i < argc)
        outfilename = argv [i++];
    else
    {
        fprintf (stderr, _("%s: destination filename missing\n"), invocation_name);
        goto error_try_help;
    }


    /* Open files and read header.  */
    init_files (strcmp (infilename, "-")  != 0 ? infilename  : 0,
                strcmp (outfilename, "-") != 0 ? outfilename : 0,
                &image_width, &image_height);


    /* Read pixel block specifications from a file.  */
    rbtree_create (&points);

    if (opt_blocks_file)
    {
        FILE *f;
        size_t len_line = 80;
        char *line = xmalloc (len_line);
        unsigned int lineno = 0;

        if ((f = fopen (arg_blocks_file, "r")) == 0)
        {
            fprintf (stderr, _("%s: %s: %s\n"), invocation_name, arg_blocks_file, strerror (errno));
            exit (1);
        }

        while (read_line (f, arg_blocks_file, &line, &len_line, &lineno))
        {
            struct point_dimdir_s new;

            if (parse_struct_point_dimdir_s (&new, line, image_width, image_height) != 0)
            {
                struct point_dimdir_s *old;
                int exists;

                if ((old = rbtree_insert (&points, &new, sizeof new, point_s_cmp, &exists)) == 0)
                    mem_alloc_failed ();
                if (exists)
                    fprintf (stderr, _("%s: %s:%u: %u,%u,%u,%u ignored: overlaps with %u,%u,%u,%u\n"),
                             invocation_name, arg_blocks_file, lineno,
                             new.p.x, new.p.y, new.p.x_size, new.p.y_size,
                             old->p.x, old->p.y, old->p.x_size, old->p.y_size);
            }
            else
            {
                fprintf (stderr, _("%s: %s:%u: invalid pixel block specification ignored\n"),
                         invocation_name, arg_blocks_file, lineno);
            }
        }

        if (fclose (f) == EOF)
        {
            fprintf (stderr, _("%s: %s: %s\n"), invocation_name, arg_blocks_file, strerror (errno));
            exit (1);
        }

        free (line);
    }


    /* Read pixel block specifications from the command line.  */
    while (i < argc)
    {
        struct point_dimdir_s new;

        if (parse_struct_point_dimdir_s (&new, argv [i], image_width, image_height) != 0)
        {
            struct point_dimdir_s *old;
            int exists;

            if ((old = rbtree_insert (&points, &new, sizeof new, point_s_cmp, &exists)) == 0)
                mem_alloc_failed ();
            if (exists)
                fprintf (stderr, _("%s: %u,%u,%u,%u ignored: overlaps with %u,%u,%u,%u\n"), invocation_name,
                         new.p.x, new.p.y, new.p.x_size, new.p.y_size,
                         old->p.x, old->p.y, old->p.x_size, old->p.y_size);
        }
        else
        {
            fprintf (stderr, _("%s: invalid pixel block specification `%s'\n"), invocation_name, argv [i]);
            goto error_try_help;
        }
        ++i;
    }


    /* Consolidate pixels.  */
    consolidate_pixels (&points, point_dimdir_s_validate_and_consolidate);


    /* Interpolate pixels in specified file.  */
    init_interpolator ();
    process_file (&points, method);

    exit (0);
}



/* Read a line from a file.  Dynamically resize buffer if necessary.
   Return pointer to the line, or a null pointer at the end of the file.  */
static char *
read_line (FILE *const f, const char *const filename, char **const buffer_ptr, size_t *const len_buffer_ptr,
           unsigned int *const lineno_ptr)
{
    size_t i = 0;
    int line_nonempty = 0, no_comment = 1;

    ++*lineno_ptr;

    while (1)
    {
#if HAVE_FGETC_UNLOCKED
        const int c = fgetc_unlocked (f);
#else
        const int c = fgetc (f);
#endif

        if (i >= *len_buffer_ptr)
            *buffer_ptr = xrealloc (*buffer_ptr, *len_buffer_ptr *= 2);

        if (c == EOF)
        {
            if (ferror (f))
            {
                fprintf (stderr, _("%s: %s: %s\n"), invocation_name, filename, strerror (errno));
                exit (1);
            }

            (*buffer_ptr) [i] = '\0';
            return line_nonempty ? *buffer_ptr : 0;
        }
        else if (c == '\n')
        {
            if (line_nonempty)
            {
                (*buffer_ptr) [i] = '\0';
                return *buffer_ptr;
            }
            else
            {
                i = 0;
                no_comment = 1;
                ++*lineno_ptr;
            }
        }
        else if (c == '#')
            no_comment = 0;
        else if (no_comment && c != '\0')
        {
            (*buffer_ptr) [i++] = c;
            if (!isspace (c))
                line_nonempty = 1;
        }
    }
}



/* Parse coordinate and size specification, and dimension and direction of
   the interpolation. Return `point' on succes, 0 otherwise.  */
static struct point_dimdir_s *
parse_struct_point_dimdir_s (struct point_dimdir_s *const point, const char *str,
                             const unsigned int image_width, const unsigned int image_height)
{
    while (isspace ((unsigned char)*str))
        ++str;

    if (*str == 'h' || *str == 'H')
        point->d = horizontal;
    else if (*str == 'v' || *str == 'V')
        point->d = vertical;
    else
    {
        point->d = twodim;
        if (*str != '2' || isdigit ((unsigned char)str [1]) || str [1] == '.' || str [1] == '%'
            || str [1] == ',' || str [1] == '/' || str [1] == ';')
            goto parse_x;
    }
    ++str;

    while (isspace ((unsigned char)*str))
        ++str;

    if (*str == ':')
        ++str;
    else if (point->d == twodim && (*str == ',' || *str == '/' || *str == ';'))
    {
        ++str;
        point->p.x = 2;
        goto parse_y;
    }
    else
        return 0;

    while (isspace ((unsigned char)*str))
        ++str;

  parse_x:
    switch (parse_number (&str, &point->p.x))
    {
      case -1:
        return 0;
      case 0:
        point->p.x = REL_TO_ABS (point->p.x, image_width);
    }

    while (isspace ((unsigned char)*str))
        ++str;

    if (*str == ',' || *str == '/' || *str == ';')
        ++str;
    else
        return 0;

  parse_y:
    while (isspace ((unsigned char)*str))
        ++str;

    switch (parse_number (&str, &point->p.y))
    {
      case -1:
        return 0;
      case 0:
        point->p.y = REL_TO_ABS (point->p.y, image_height);
    }

    while (isspace ((unsigned char)*str))
        ++str;

    if (*str == ',' || *str == '/' || *str == ';')
        ++str;
    else if (*str == '\0')
    {
        point->p.x_size = 1;
        point->p.y_size = 1;
        return point;
    }
    else
        return 0;

    while (isspace ((unsigned char)*str))
        ++str;

    switch (parse_number (&str, &point->p.x_size))
    {
      case -1:
        return 0;
      case 0:
        point->p.x_size = REL_TO_ABS (point->p.x_size, image_width);
    }
    if (point->p.x_size == 0)
        return 0;

    while (isspace ((unsigned char)*str))
        ++str;

    if (*str == ',' || *str == '/' || *str == ';')
        ++str;
    else if (*str == '\0')
    {
        point->p.y_size = point->p.x_size;
        return point;
    }
    else
        return 0;

    while (isspace ((unsigned char)*str))
        ++str;

    switch (parse_number (&str, &point->p.y_size))
    {
      case -1:
        return 0;
      case 0:
        point->p.y_size = REL_TO_ABS (point->p.y_size, image_height);
    }
    if (point->p.y_size == 0)
        return 0;

    while (isspace ((unsigned char)*str))
        ++str;

    return *str == '\0' ? point : 0;
}



/* Validate and consolidate two point structures.  */
static int
point_dimdir_s_validate_and_consolidate (void *const a, const void *const b)
{
    struct point_dimdir_s *const point_a = a;
    const struct point_dimdir_s *const point_b = b;

    if (point_a->d != point_b->d)
    {
        fprintf (stderr, _("%s: incompatible adjacent points %u,%u,%u,%u (%s) and %u,%u,%u,%u (%s) ignored\n"),
                 invocation_name, point_b->p.x, point_b->p.y, point_b->p.x_size, point_b->p.y_size,
                 point_b->d == vertical ? _("vertical") : point_b->d == horizontal ? _("horizontal") : _("2-dim"),
                 point_a->p.x, point_a->p.y, point_a->p.x_size, point_a->p.y_size,
                 point_a->d == vertical ? _("vertical") : point_a->d == horizontal ? _("horizontal") : _("2-dim"));
        return 0;
    }

    return 1;
}
